presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java

```diff
@@ -13,33 +13,65 @@
  */
 package com.facebook.presto.hive;
 
+import com.facebook.presto.common.type.Type;
 import com.facebook.presto.spi.connector.ConnectorPartitioningHandle;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 import java.util.List;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.OptionalInt;
 
+import static com.facebook.presto.hive.HiveBucketProperty.BucketingType.HIVETYPE;
+import static com.facebook.presto.hive.HiveBucketProperty.BucketingType.PRESTOTYPE;
+import static com.google.common.base.Preconditions.checkArgument;
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
 public class HivePartitioningHandle
         implements ConnectorPartitioningHandle
 {
     private final int bucketCount;
-    private final List<HiveType> hiveTypes;
+    private final Optional<List<HiveType>> hiveTypes;
+    private final Optional<List<Type>> types;
     private final OptionalInt maxCompatibleBucketCount;
+    private final HiveBucketProperty.BucketingType bucketingType;
 
+    public static HivePartitioningHandle createPartitioningHandleWithHiveType(
+            int bucketCount,
+            List<HiveType> hiveTypes,
+            OptionalInt maxCompatibleBucketCount)
+    {
+        return new HivePartitioningHandle(bucketCount, maxCompatibleBucketCount, Optional.of(hiveTypes), Optional.empty());
+    }
+
+    public static HivePartitioningHandle createPartitioningHandleWithPrestoType(
+            int bucketCount,
+            List<Type> types,
+            OptionalInt maxCompatibleBucketCount)
+    {
+        return new HivePartitioningHandle(bucketCount, maxCompatibleBucketCount, Optional.empty(), Optional.of(types));
+    }
+
+    /**
+     *  for json serialization only, don't use this method to create HivePartitioningHandle
+     */
     @JsonCreator
     public HivePartitioningHandle(
             @JsonProperty("bucketCount") int bucketCount,
-            @JsonProperty("hiveTypes") List<HiveType> hiveTypes,
-            @JsonProperty("maxCompatibleBucketCount") OptionalInt maxCompatibleBucketCount)
+            @JsonProperty("maxCompatibleBucketCount") OptionalInt maxCompatibleBucketCount,
+            @JsonProperty("hiveTypes") Optional<List<HiveType>> hiveTypes,
+            @JsonProperty("types") Optional<List<Type>> types)
```

Comment:
Let's accept `BucketingType` as a parameter and check if `types` or `hiveTypes` are present based on the `BucketingType`.